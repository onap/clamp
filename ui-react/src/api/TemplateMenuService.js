/*-
 * ============LICENSE_START=======================================================
 * ONAP CLAMP
 * ================================================================================
 * Copyright (C) 2019 AT&T Intellectual Property. All rights reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */

export default class TemplateMenuService {
  static getToscaPolicyModels() {
    return fetch('restservices/clds/v2/loop/tosca/models', { method: 'GET', credentials: 'same-origin' })
      .then(function (response) {
        console.debug("getToscaPolicyModels response received: ", response.status);
        if (response.ok) {
          return response.json();
        } else {
          console.error("getToscaPolicyModels query failed");
          return {};
        }
      })
      .catch(function (error) {
        console.error("getToscaPolicyModels error received", error);
        return {};
      });
  }

	static getToscaPolicyModelYaml(policyModelType) {
		return fetch('/restservices/clds/v2/loop/tosca/models/yaml/' + policyModelType, {
			method: 'GET',
			credentials: 'same-origin'
		})
			.then(function (response) {
				console.debug("getToscaPolicyModelYaml response received: ", response.status);
				if (response.ok) {
					return response.json();
				} else {
					console.error("getToscaPolicyModelYaml query failed");
					return "";
				}
			})
			.catch(function (error) {
				console.error("getToscaPolicyModelYaml error received", error);
				return "";
			});
	}

  static uploadToscaPolicyModal(policyModelType, jsonData) {
    return fetch('/restservices/clds/v2/loop/tosca/models/' + policyModelType, {
        method: 'PUT',
        credentials: 'same-origin',
        headers: {
          "Content-Type": "a",
        },
        body: JSON.stringify(jsonData)
      })
      .then(function(response) {
        console.debug("uploadToscaPolicyModal response received: ", response.status);
        if (response.ok) {
          var message = {
            status: response.status,
            message: 'Tosca Policy Model successfully uploaded'
          };
          return message;
        } else {
          console.error("uploadToscaModel failed");
          return response.text();
        }
      })
      .catch(function(error) {
        console.error("uploadToscaModal error received", error);
        return "";
      });
  }

  static getBlueprintMicroServiceTemplates() {
    return fetch('restservices/clds/v2/loop/templates', { method: 'GET', credentials: 'same-origin', })
      .then(function (response) {
        console.debug("getBlueprintMicroServiceTemplates response received: ", response.status);
        if (response.ok) {
          return response.json();
        } else {
          console.error("getBlueprintMicroServiceTemplates query failed");
          return {};
        }
      })
      .catch(function (error) {
        console.error("getBlueprintMicroServiceTemplates error received", error);
        return {};
      });
  }
}
