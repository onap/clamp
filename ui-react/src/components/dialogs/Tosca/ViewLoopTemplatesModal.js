/*-
 * ============LICENSE_START=======================================================
 * ONAP CLAMP
 * ================================================================================
 * Copyright (C) 2019 AT&T Intellectual Property. All rights
 *                             reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */

import React, { forwardRef } from 'react'
import Button from 'react-bootstrap/Button';
import Modal from 'react-bootstrap/Modal';
import styled from 'styled-components';
import TemplateMenuService from '../../../api/TemplateMenuService';
import ArrowUpward from '@material-ui/icons/ArrowUpward';
import ChevronLeft from '@material-ui/icons/ChevronLeft';
import ChevronRight from '@material-ui/icons/ChevronRight';
import Clear from '@material-ui/icons/Clear';
import FirstPage from '@material-ui/icons/FirstPage';
import LastPage from '@material-ui/icons/LastPage';
import Search from '@material-ui/icons/Search';
import MaterialTable from "material-table";

const ModalStyled = styled(Modal)`
	background-color: transparent;
`
const LoopViewSvgDivStyled = styled.div`
overflow: hidden;
background-color: ${props => (props.theme.loopViewerBackgroundColor)};
border: 1px solid;
border-color: ${props => (props.theme.loopViewerHeaderColor)};
margin-left: auto;
margin-right:auto;
text-align: center;
`
const cellStyle = { border: '1px solid black' };
const headerStyle = { backgroundColor: '#ddd',	border: '2px solid black'	};
const rowHeaderStyle = {backgroundColor:'#ddd',  fontSize: '15pt', text: 'bold', border: '1px solid black'};

export default class ViewLoopTemplatesModal extends React.Component {
  state = {
    show: true,
		content: 'Please select Blue print template to view the details',
		selectedRow: -1,
		loopTemplateData: [],
		svgRepresentation: [],
		loopTemplateColumnsDefinition: [
			{ title: "#", field: "index", render: rowData => rowData.tableData.id + 1,
				cellStyle: cellStyle,
				headerStyle: headerStyle
			},
			{ title: "Template Name", field: "name",
				cellStyle: cellStyle,
				headerStyle: headerStyle
			},
			{ title: "Maximum Instances Allowed", field: "maximumInstancesAllowed",
				cellStyle: cellStyle,
				headerStyle: headerStyle
			},
			{ title: "Loop Type Allowed ", field: "allowedLoopType",
				cellStyle: cellStyle,
				headerStyle: headerStyle
			},
			{ title: "Unique Blueprint", field: "uniqueBlueprint", lookup: {false: 'No', true: 'Yes'},
				cellStyle: cellStyle,
				headerStyle: headerStyle
			},
			{ title: "Modified By", field: "updatedBy",
				cellStyle: cellStyle,
				headerStyle: headerStyle
			},
			{ title: "Modified Date", field: "updatedDate",
				cellStyle: cellStyle,
				headerStyle: headerStyle
			}
		],
		tableIcons: {
			FirstPage: forwardRef((props, ref) => <FirstPage {...props} ref={ref} />),
			LastPage: forwardRef((props, ref) => <LastPage {...props} ref={ref} />),
			NextPage: forwardRef((props, ref) => <ChevronRight {...props} ref={ref} />),
			PreviousPage: forwardRef((props, ref) => <ChevronLeft {...props} ref={ref} />),
			ResetSearch: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
			Search: forwardRef((props, ref) => <Search {...props} ref={ref} />),
			SortArrow: forwardRef((props, ref) => <ArrowUpward {...props} ref={ref} />)
		}
	};

	constructor(props, context) {
		super(props, context);
		this.handleClose = this.handleClose.bind(this);
		this.getLoopTemplates = this.getLoopTemplates.bind(this);
		this.getSVGRepresentation = this.getSVGRepresentation.bind(this);
		this.handleYamlContent = this.handleYamlContent.bind(this);
	}

	componentWillMount() {
		this.getLoopTemplates();
	}

	getLoopTemplates() {
		TemplateMenuService.getLoopTemplates().then(loopTemplateData => {
			var TemplNames = loopTemplateData;
			for(var item in TemplNames) {
				if(TemplNames[item].maximumInstancesAllowed === -1) {
					TemplNames[item].maximumInstancesAllowed = 'No Limit';
				}
			}
		this.setState({ loopTemplateData: TemplNames })
		});
	}

	getSVGRepresentation(templateName) {
		TemplateMenuService.getSVGRepresentation(templateName).then(svgRepresentation => {
				this.setState({ content: svgRepresentation});
			});
		}

	handleYamlContent = event => {
		this.setState({
			content: event.target.value
		});
	}

	handleClose() {
		this.setState({ show: false });
		this.props.history.push('/')
	}

	render() {
    return (
      <ModalStyled size="xl" show={this.state.show} onHide={this.handleClose}>
        <Modal.Header closeButton>
				</Modal.Header>
				<Modal.Body>
          <MaterialTable
            title={"View Blueprint MicroService Templates"}
            data={this.state.loopTemplateData}
					  columns={this.state.loopTemplateColumnsDefinition}
					  icons={this.state.tableIcons}
					  onRowClick={(event, rowData) => {this.getSVGRepresentation(rowData.name);this.setState({selectedRow: rowData.tableData.id})}}
					  options={{
					  headerStyle:rowHeaderStyle,
					  rowStyle: rowData => ({
					  backgroundColor: (this.state.selectedRow !== -1 && this.state.selectedRow === rowData.tableData.id) ? '#EEE' : '#FFF'
            })
          }}
          />
					<LoopViewSvgDivStyled
						dangerouslySetInnerHTML={{ __html: this.state.content }} onChange={this.handleYamlContent}>
          </LoopViewSvgDivStyled>
        </Modal.Body>
        <Modal.Footer>
          <Button variant="secondary" onClick={this.handleClose}>Close</Button>
        </Modal.Footer>
      </ModalStyled>
      );
    }
  }
