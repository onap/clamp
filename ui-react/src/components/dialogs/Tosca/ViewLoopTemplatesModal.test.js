/*-
 * ============LICENSE_START=======================================================
 * ONAP CLAMP
 * ================================================================================
 * Copyright (C) 2019 AT&T Intellectual Property. All rights
 *                             reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */
import React from 'react';
import { shallow } from 'enzyme';
import ViewLoopTemplatesModal from './ViewLoopTemplatesModal';
import { mount } from 'enzyme';
import TemplateMenuService from '../../../api/TemplateMenuService'

describe('Verify ViewLoopTemplatesModal', () => {
	beforeEach(() => {
		fetch.resetMocks();
		}
	);

	it('Test API Successful', () => {
		fetch.mockImplementationOnce(() => {
			return Promise.resolve({
				ok: true,
				status: 200,
				json: () => {
					return Promise.resolve({
						"index": "1",
						"name": "TOSCA_MTCA",
						"maximumInstancesAllowed": "1",
						"allowedLoopType" : "OPEN",
						"uniqueBlueprint":"No",
						"updatedBy":"admin",
						"updatedDate":"2019-09-06 19:09:42"
					});
				}
			});
		});
		const component = shallow(<ViewLoopTemplatesModal/>);
	});

	it('Test API Exception', () => {
		fetch.mockImplementationOnce(() => {
			return Promise.resolve({
				ok: false,
				status: 500,
				json: () => {
					return Promise.resolve({
						"index": "1",
						"name": "TOSCA_MTCA",
						"maximumInstancesAllowed": "1",
						"allowedLoopType" : "OPEN",
						"uniqueBlueprint":"No",
						"updatedBy":"admin",
						"updatedDate":"2019-09-06 19:09:42"
					});
				}
			});
		});
		const component = shallow(<ViewLoopTemplatesModal/>);
	});

	it('Test API Rejection', () => {
		const myMockFunc  = fetch.mockImplementationOnce(() => Promise.reject('error'));
		setTimeout( () => myMockFunc().catch(e => {
			console.log(e);
		}),
		100
		);
		new Promise(resolve => setTimeout(resolve, 200));
		const component = shallow(<ViewLoopTemplatesModal/>);
		expect(myMockFunc.mock.calls.length).toBe(1);
	});

	it('Test the BP template render method', () => {
		fetch.mockImplementationOnce(() => {
			return Promise.resolve({
				ok: true,
				status: 200,
				json: () => {
					return Promise.resolve({
						"index": "1",
						"name": "TOSCA_MTCA",
						"maximumInstancesAllowed": "1",
						"allowedLoopType" : "OPEN",
						"uniqueBlueprint":"No",
						"updatedBy":"admin",
						"updatedDate":"2019-09-06 19:09:42"
					});
				}
			});
		});
		const component = shallow(<ViewLoopTemplatesModal/>);
		component.setState({ loopTemplateData: {
			"index": "1",
			"name": "TOSCA_MTCA",
			"maximumInstancesAllowed": "1",
			"allowedLoopType" : "OPEN",
			"uniqueBlueprint":"No",
			"updatedBy":"admin",
			"updatedDate":"2019-09-06 19:09:42"
		  }
		});
		expect(component).toMatchSnapshot();
	});

	it('Test Table icons', () => {
			fetch.mockImplementationOnce(() => {
				return Promise.resolve({
					ok: true,
					status: 200,
					json: () => {
						return Promise.resolve({
							"index": "1",
							"name": "TOSCA_MTCA",
							"maximumInstancesAllowed": "1",
							"allowedLoopType" : "OPEN",
							"uniqueBlueprint":"No",
							"updatedBy":"admin",
							"updatedDate":"2019-09-06 19:09:42"
						});
					}
				});
			});
			const component = mount(<ViewLoopTemplatesModal/>);
			expect(component.find('[className="MuiSelect-icon MuiTablePagination-selectIcon"]')).toBeTruthy();
		});

		it('Test handleYamlContent', () => {
			fetch.mockImplementationOnce(() => {
				return Promise.resolve({
					ok: true,
					status: 200,
					json: () => {
						return Promise.resolve({
							"index": "1",
							"name": "TOSCA_MTCA",
							"maximumInstancesAllowed": "1",
							"allowedLoopType" : "OPEN",
							"uniqueBlueprint":"No",
							"updatedBy":"admin",
							"updatedDate":"2019-09-06 19:09:42"
						});
					}
				});
			});
			const yamlContent = 'Please select Blue print template to view the details';
			const component = shallow(<ViewLoopTemplatesModal/>);
			expect(component.find('[className="sc-kAzzGY fzcWwp"]')).toBeTruthy();
			expect(component.state('content')).toEqual(yamlContent);
		});

		it('Test handleClose', () => {
			fetch.mockImplementationOnce(() => {
				return Promise.resolve({
					ok: true,
					status: 200,
					json: () => {
						return Promise.resolve({
							"index": "1",
							"name": "TOSCA_MTCA",
							"maximumInstancesAllowed": "1",
							"allowedLoopType" : "OPEN",
							"uniqueBlueprint":"No",
							"updatedBy":"admin",
							"updatedDate":"2019-09-06 19:09:42"
						});
					}
				});
			});
			const historyMock = { push: jest.fn() };
			const handleClose = jest.spyOn(ViewLoopTemplatesModal.prototype,'handleClose');
			const component = shallow(<ViewLoopTemplatesModal history={historyMock} />)
			component.find('[variant="secondary"]').prop('onClick')();
			expect(handleClose).toHaveBeenCalledTimes(1);
			expect(component.state('show')).toEqual(false);
			expect(historyMock.push.mock.calls[0]).toEqual([ '/']);
			handleClose.mockClear();
		});

		test('Test getSVGRepresentation', async () => {
			const historyMock = { push: jest.fn() };
			TemplateMenuService.getLoopTemplates = jest.fn().mockImplementation(() => {
				return Promise.resolve({loopTemplateData: "TOSCA_MTCA"});
			});
			TemplateMenuService.getSVGRepresentation = jest.fn().mockImplementation(() => {
				return Promise.resolve("<svg><text test</text></svg>");
			});
			const flushPromises = () => new Promise(setImmediate);
			const component = shallow(<ViewLoopTemplatesModal history={historyMock}/>)
			const instance = component.instance();
			instance.getSVGRepresentation("TOSCA_MTCA");
			await flushPromises();
			expect(component.state('content')).toEqual("<svg><text test</text></svg>");
		})
	});
