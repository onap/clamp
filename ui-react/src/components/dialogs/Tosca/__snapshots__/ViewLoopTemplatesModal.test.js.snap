// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify ViewLoopTemplatesModal Test the BP template render method 1`] = `
<Styled(Bootstrap(Modal))
  onHide={[Function]}
  show={true}
  size="xl"
>
  <ModalHeader
    closeButton={true}
    closeLabel="Close"
  />
  <ModalBody>
    <WithStyles(Component)
      columns={
        Array [
          Object {
            "cellStyle": Object {
              "border": "1px solid black",
            },
            "field": "index",
            "headerStyle": Object {
              "backgroundColor": "#ddd",
              "border": "2px solid black",
            },
            "render": [Function],
            "title": "#",
          },
          Object {
            "cellStyle": Object {
              "border": "1px solid black",
            },
            "field": "name",
            "headerStyle": Object {
              "backgroundColor": "#ddd",
              "border": "2px solid black",
            },
            "title": "Template Name",
          },
          Object {
            "cellStyle": Object {
              "border": "1px solid black",
            },
            "field": "maximumInstancesAllowed",
            "headerStyle": Object {
              "backgroundColor": "#ddd",
              "border": "2px solid black",
            },
            "title": "Maximum Instances Allowed",
          },
          Object {
            "cellStyle": Object {
              "border": "1px solid black",
            },
            "field": "allowedLoopType",
            "headerStyle": Object {
              "backgroundColor": "#ddd",
              "border": "2px solid black",
            },
            "title": "Loop Type Allowed ",
          },
          Object {
            "cellStyle": Object {
              "border": "1px solid black",
            },
            "field": "uniqueBlueprint",
            "headerStyle": Object {
              "backgroundColor": "#ddd",
              "border": "2px solid black",
            },
            "lookup": Object {
              "false": "No",
              "true": "Yes",
            },
            "title": "Unique Blueprint",
          },
          Object {
            "cellStyle": Object {
              "border": "1px solid black",
            },
            "field": "updatedBy",
            "headerStyle": Object {
              "backgroundColor": "#ddd",
              "border": "2px solid black",
            },
            "title": "Modified By",
          },
          Object {
            "cellStyle": Object {
              "border": "1px solid black",
            },
            "field": "updatedDate",
            "headerStyle": Object {
              "backgroundColor": "#ddd",
              "border": "2px solid black",
            },
            "title": "Modified Date",
          },
        ]
      }
      data={
        Object {
          "allowedLoopType": "OPEN",
          "index": "1",
          "maximumInstancesAllowed": "1",
          "name": "TOSCA_MTCA",
          "uniqueBlueprint": "No",
          "updatedBy": "admin",
          "updatedDate": "2019-09-06 19:09:42",
        }
      }
      icons={
        Object {
          "FirstPage": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
          "LastPage": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
          "NextPage": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
          "PreviousPage": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
          "ResetSearch": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
          "Search": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
          "SortArrow": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        }
      }
      onRowClick={[Function]}
      options={
        Object {
          "headerStyle": Object {
            "backgroundColor": "#ddd",
            "border": "1px solid black",
            "fontSize": "15pt",
            "text": "bold",
          },
          "rowStyle": [Function],
        }
      }
      title="View Blueprint MicroService Templates"
    />
    <styled.div
      dangerouslySetInnerHTML={
        Object {
          "__html": "Please select Blue print template to view the details",
        }
      }
      onChange={[Function]}
    />
  </ModalBody>
  <ModalFooter>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      type="button"
      variant="secondary"
    >
      Close
    </Button>
  </ModalFooter>
</Styled(Bootstrap(Modal))>
`;
